### PR TITLE
fixed tearDowns and rewrite `payLoan` to `takeLoan`

### PR DESCRIPTION
**Pull request by ith-harvey**
Solved not one... but TWO `FIXME` 's in the tests. The cause was a bug in the situations in which "loans" which turned out to actually be active auctions had the `loans.update` method called on them. This resulted in live auctions being turned back into active loans :scream: 

This PR fixes `tearDown` for two tests:
* `testMergeOrRemoveCollateral`
* `testSettleOnAuctionKicked72HoursAgoAndPartiallyTaken`

Changes:
`loans.update` now accepts an additional argument `inAuction_` which blocks an auction from be added or removed from the loan heap

Challenges:
Not a big fan of how I implemented the change as I don't think its elegant. Would love feeback to figure out a cleaner way to perform this check or better yet remove the check alltogether and refactor how / when / what loans `loans.update` is called on.

Here is another admittedly unelegant idea I had for where we want to call `loans.update`:
       if (Auctions.isActive(auctions, borrowerAddress_)) {
            loans.borrowers[borrowerAddress_] = borrower_;
        } else {
            loans.update(borrowerAddress_, borrower_, loanId);
        }
        
Contract size & gas before:
```
  ERC721Pool               -  23,404B  (95.23%)
  ERC20Pool                -  20,935B  (85.18%)
  PositionManager          -  17,916B  (72.90%)
  Auctions                 -  13,544B  (55.11%)
  PoolInfoUtils            -  11,574B  (47.09%)
  LenderActions            -   9,961B  (40.53%)
  PoolCommons              -   7,400B  (30.11%)
  ERC721                   -   4,421B  (17.99%)
  ERC721PoolFactory        -   2,876B  (11.70%)
  ERC20PoolFactory         -   2,194B  (8.93%)
  ERC20                    -   2,140B  (8.71%)
```

```
| src/erc721/ERC721Pool.sol:ERC721Pool contract |                 |        |        |         |         |
|-----------------------------------------------|-----------------|--------|--------|---------|---------|
| Deployment Cost                               | Deployment Size |        |        |         |         |
| 4708395                                       | 23442           |        |        |         |         |
| Function Name                                 | min             | avg    | median | max     | # calls |
| addCollateral                                 | 232712          | 364194 | 266333 | 593538  | 3       |
| addQuoteToken                                 | 72412           | 193658 | 150842 | 459400  | 177     |
| approveLpOwnership                            | 20630           | 21942  | 22730  | 22730   | 8       |
| auctionInfo                                   | 1168            | 2668   | 1168   | 9168    | 16      |
| borrowerInfo                                  | 844             | 1162   | 844    | 6844    | 226     |
| borrowerTokenIds                              | 892             | 1551   | 892    | 4892    | 94      |
| bucketInfo                                    | 6997            | 8980   | 8120   | 26997   | 202     |
| bucketTake                                    | 59421           | 68066  | 65018  | 413601  | 100     |
| bucketTokenIds                                | 709             | 709    | 709    | 709     | 7       |
| collateralAddress                             | 503             | 503    | 503    | 503     | 8       |
| debtInfo                                      | 5088            | 5878   | 5088   | 16008   | 260     |
| depositIndex                                  | 1043            | 15534  | 12596  | 68559   | 154     |
| depositSize                                   | 816             | 900    | 816    | 4816    | 95      |
| depositUtilization                            | 1114            | 35430  | 29156  | 63221   | 16      |
| drawDebt                                      | 27825           | 214538 | 209673 | 387792  | 101     |
| emasInfo                                      | 540             | 1716   | 540    | 4540    | 17      |
| flashFee                                      | 619             | 629    | 619    | 650     | 3       |
| flashLoan                                     | 23156           | 38745  | 30063  | 73622   | 6       |
| inflatorInfo                                  | 443             | 533    | 443    | 2443    | 221     |
| initialize                                    | 3087            | 168119 | 181565 | 293595  | 75      |
| interestRateInfo                              | 388             | 498    | 388    | 2388    | 199     |
| isSubset                                      | 422             | 422    | 422    | 422     | 2       |
| kick                                          | 726             | 287058 | 374942 | 455690  | 7       |
| kickerInfo                                    | 752             | 1447   | 752    | 4752    | 23      |
| lenderInfo                                    | 1007            | 1955   | 1274   | 7274    | 172     |
| loansInfo                                     | 1839            | 2490   | 2537   | 5839    | 101     |
| maxFlashLoan                                  | 6246            | 6246   | 6246   | 6246    | 1       |
| mergeOrRemoveCollateral                       | 3409            | 546992 | 56196  | 1581371 | 3       |
| moveQuoteToken                                | 62797           | 62797  | 62797  | 62797   | 1       |
| onERC721Received                              | 1050            | 1050   | 1050   | 1050    | 112     |
| pledgedCollateral                             | 385             | 497    | 385    | 2385    | 71      |
| poolType                                      | 402             | 402    | 402    | 402     | 2       |
| quoteTokenAddress                             | 458             | 458    | 458    | 458     | 143     |
| quoteTokenScale                               | 432             | 432    | 432    | 432     | 2       |
| removeCollateral                              | 6362            | 56573  | 42733  | 138493  | 9       |
| removeQuoteToken                              | 43978           | 54289  | 48270  | 185206  | 59      |
| repayDebt                                     | 24123           | 117929 | 71668  | 527017  | 75      |
| reservesInfo                                  | 629             | 1669   | 629    | 6629    | 50      |
| settle                                        | 83194           | 161775 | 161775 | 240356  | 2       |
| startClaimableReserveAuction                  | 7122            | 49504  | 74650  | 76650   | 5       |
| take                                          | 72605           | 354899 | 356513 | 580507  | 5       |
| takeReserves                                  | 1120            | 35866  | 32508  | 72761   | 6       |
| tokenIdsAllowed                               | 2495            | 2495   | 2495   | 2495    | 5       |
| transferLPTokens                              | 10496           | 66540  | 78310  | 110816  | 3       |
| withdrawBonds                                 | 4392            | 4392   | 4392   | 4392    | 1       |

```

Contract size & gas after changes:
```
============ Deployment Bytecode Sizes ============
  ERC721Pool               -  23,445B  (95.39%)
  ERC20Pool                -  20,976B  (85.35%)
  PositionManager          -  17,916B  (72.90%)
  Auctions                 -  13,544B  (55.11%)
  PoolInfoUtils            -  11,574B  (47.09%)
  LenderActions            -   9,961B  (40.53%)
  PoolCommons              -   7,400B  (30.11%)
  ERC721                   -   4,421B  (17.99%)
  ERC721PoolFactory        -   2,876B  (11.70%)
  ERC20PoolFactory         -   2,194B  (8.93%)
  ```
  ```
  | src/erc721/ERC721Pool.sol:ERC721Pool contract |                 |        |        |         |         |
|-----------------------------------------------|-----------------|--------|--------|---------|---------|
| Deployment Cost                               | Deployment Size |        |        |         |         |
| 4716604                                       | 23483           |        |        |         |         |
| Function Name                                 | min             | avg    | median | max     | # calls |
| addCollateral                                 | 232712          | 364194 | 266333 | 593538  | 3       |
| addQuoteToken                                 | 72412           | 193658 | 150842 | 459400  | 177     |
| approveLpOwnership                            | 20630           | 21942  | 22730  | 22730   | 8       |
| auctionInfo                                   | 1168            | 2579   | 1168   | 9168    | 17      |
| borrowerInfo                                  | 844             | 1158   | 844    | 6844    | 229     |
| borrowerTokenIds                              | 892             | 1531   | 892    | 4892    | 97      |
| bucketInfo                                    | 5859            | 9403   | 8422   | 26997   | 307     |
| bucketTake                                    | 58071           | 66010  | 63667  | 341632  | 100     |
| bucketTokenIds                                | 709             | 709    | 709    | 709     | 7       |
| collateralAddress                             | 503             | 503    | 503    | 503     | 8       |
| debtInfo                                      | 5088            | 5850   | 5088   | 16008   | 271     |
| depositIndex                                  | 1043            | 15035  | 12559  | 68559   | 163     |
| depositSize                                   | 816             | 896    | 816    | 4816    | 99      |
| depositUtilization                            | 1114            | 36122  | 30178  | 63464   | 19      |
| drawDebt                                      | 28095           | 214830 | 209966 | 388085  | 101     |
| emasInfo                                      | 540             | 1740   | 540    | 4540    | 20      |
| flashFee                                      | 619             | 629    | 619    | 650     | 3       |
| flashLoan                                     | 23156           | 38745  | 30063  | 73622   | 6       |
| inflatorInfo                                  | 443             | 528    | 443    | 2443    | 233     |
| initialize                                    | 3087            | 168119 | 181565 | 293595  | 75      |
| interestRateInfo                              | 388             | 493    | 388    | 2388    | 208     |
| isSubset                                      | 422             | 422    | 422    | 422     | 2       |
| kick                                          | 726             | 287058 | 374942 | 455690  | 7       |
| kickerInfo                                    | 752             | 1418   | 752    | 4752    | 24      |
| lenderInfo                                    | 1007            | 1708   | 1274   | 7274    | 278     |
| loansInfo                                     | 1839            | 2431   | 2537   | 5839    | 111     |
| maxFlashLoan                                  | 6246            | 6246   | 6246   | 6246    | 1       |
| mergeOrRemoveCollateral                       | 3409            | 546992 | 56196  | 1581371 | 3       |
| moveQuoteToken                                | 62797           | 62797  | 62797  | 62797   | 1       |
| onERC721Received                              | 1050            | 1050   | 1050   | 1050    | 112     |
| pledgedCollateral                             | 385             | 487    | 385    | 2385    | 78      |
| poolType                                      | 402             | 402    | 402    | 402     | 2       |
| quoteTokenAddress                             | 458             | 458    | 458    | 458     | 145     |
| quoteTokenScale                               | 432             | 432    | 432    | 432     | 2       |
| removeCollateral                              | 6362            | 56573  | 42733  | 138493  | 9       |
| removeQuoteToken                              | 43978           | 54018  | 48181  | 185206  | 61      |
| repayDebt                                     | 24123           | 117053 | 72136  | 527309  | 76      |
| reservesInfo                                  | 629             | 1648   | 629    | 6629    | 51      |
| settle                                        | 83194           | 141137 | 141137 | 199080  | 2       |
| startClaimableReserveAuction                  | 7122            | 49504  | 74650  | 76650   | 5       |
| take                                          | 71332           | 354820 | 356747 | 580711  | 5       |
| takeReserves                                  | 1120            | 35866  | 32508  | 72761   | 6       |
| tokenIdsAllowed                               | 2495            | 2495   | 2495   | 2495    | 5       |
| transferLPTokens                              | 10496           | 66540  | 78310  | 110816  | 3       |
| withdrawBonds                                 | 4392            | 4392   | 4392   | 4392    | 1       |

  ```

### CODE REVIEW

File: tests/forge/ERC20Pool/ERC20PoolLiquidationsSettle.t.sol [hunk 1]
```diff
@@ -420,15 +442,17 @@ contract ERC20PoolLiquidationsSettleTest is ERC20HelperContract {
...
+
         _assertBucket(
             {
                 index:        _i9_72,
                 lpBalance:    11_000 * 1e27,
                 collateral:   0,
-                deposit:      8_775.731868287982080143 * 1e18,
-                exchangeRate: 0.797793806207998370922090909 * 1e27
+                deposit:      8_775.721258848371782184 * 1e18,
+                exchangeRate: 0.797792841713488343834909090 * 1e27
```
Comment (ith-harvey) [inline tests/forge/ERC20Pool/ERC20PoolLiquidationsSettle.t.sol:451]: Interestingly these values were changed when we didn't add a undercollateralized loan back to the pool in the settle call above.

File: tests/forge/ERC721Pool/ERC721DSTestPlus.sol [hunk 2]
```diff
@@ -469,11 +449,39 @@ abstract contract ERC721DSTestPlus is DSTestPlus, IERC721PoolEvents {
         uint256[] memory removeCollateralAtIndex,
         uint256 collateralMerged,
         uint256 toIndexLps
-    ) internal virtual {
+    ) internal {
```
Comment (ith-harvey) [inline tests/forge/ERC721Pool/ERC721DSTestPlus.sol:403]: This is fine as `mergeOrRemoveCollateral` is only called by ERC721 pool

File: tests/forge/utils/DSTestPlus.sol [hunk 3]
```diff
@@ -173,9 +173,6 @@ abstract contract DSTestPlus is Test, IPoolEvents {
     ) internal virtual {
         changePrank(from);
         _pool.bucketTake(borrower, true, index);
-
-        // Add for tearDown
-        bucketsUsed.add(index);
```
Comment (EdNoepel) [inline tests/forge/utils/DSTestPlus.sol:178]: `depositTake` awards LPB in the bucket specified by the caller.  Why would we exclude that from the teardown?

File: tests/forge/ERC721Pool/ERC721DSTestPlus.sol [hunk 4]
```diff
@@ -28,7 +28,6 @@ abstract contract ERC721DSTestPlus is DSTestPlus, IERC721PoolEvents {
     Token              internal _quote;
     ERC20              internal _ajnaToken;
 
-    mapping(address => EnumerableSet.UintSet) borrowerPlegedNFTIds;
```
Comment (EdNoepel) [inline tests/forge/ERC721Pool/ERC721DSTestPlus.sol:31]: Glad you were able to eliminate this.

File: src/libraries/external/PoolCommons.sol [hunk 5]
```diff
@@ -139,6 +139,7 @@ library PoolCommons {
                 htpIndex,
                 Maths.wdiv(newInterest, depositAboveHtp) + Maths.WAD // lender factor
             );
+
```
Comment (EdNoepel) [inline src/libraries/external/PoolCommons.sol]: nit: why blank line here?

**Review by EdNoepel — COMMENTED** [review]
I agree it doesn't taste great passing `inAuction` into `Loans.update`, but the other options don't seem any better:
 - Adding conditionals to `Pool`, or implementing a wrapper function, would be more expensive and difficult to maintain.
 - Passing the entire `AuctionsState` into `Loans.update` would entangle concerns.
 
 Commented regarding a teardown concern.

File: tests/forge/utils/DSTestPlus.sol (hunk 3 above)
Comment (ith-harvey) [reply tests/forge/utils/DSTestPlus.sol:178]: Caller of `depositTake` gets nothing... they are not directly inscentivized to call the method. The thought is they may be someone who has deposited in the bucket that they are calling deposit take on see -> https://github.com/ajna-finance/contracts/blob/develop/src/libraries/external/Auctions.sol#L916

Comment (ith-harvey) [reply tests/forge/utils/DSTestPlus.sol:178]: in `arbTake` they are inscentivized.

File: src/libraries/external/PoolCommons.sol (hunk 5 above)
Comment (ith-harvey) [reply src/libraries/external/PoolCommons.sol]: accident, removed

File: tests/forge/utils/DSTestPlus.sol (hunk 3 above)
Comment (EdNoepel) [reply tests/forge/utils/DSTestPlus.sol:178]: But the kicker could still be awarded LP; see line 933.

Comment (EdNoepel) [reply tests/forge/utils/DSTestPlus.sol:178]: On second thought, the bucket would need liquidity before any `bucketTake` could be called on it, so this is superfluous.

**Review by EdNoepel — APPROVED** [review]

File: tests/forge/utils/DSTestPlus.sol (hunk 3 above)
Comment (ith-harvey) [reply tests/forge/utils/DSTestPlus.sol:178]: Pulled from V10 spec:
``` 
Note the caller of this method does not receive anything directly from the call. The caller may be someone with a deposit in the bucket (specified in the arguments) or may be the kicker who wants LPB in a bucket.
```

**Pull request merged**